### ISSUE TITLE
05b3 lock ups after a little more than an hour of flight.

### ISSUE THREAD

**Issue by braddward** [cyoung/stratux#180]
I went up yesterday and today and had the same thing happen after a little over an hour of flight on 05b3.

The stratux appeared to be locked up.  I power cycled and it came back online without a problem.

Here's a link to the logs:  https://app.box.com/s/xrcwvnrfu4kqbw1evgugmvanmto0j0s3

Did the fixes to set the system date/time from the GPS go in after this release?

**Comment by cyoung** [comment]
Those fixes are not in v0.5b3.

What app were you using, which GPS, how many SDRs?

Was the app receiving heartbeats after lockup? In FF you you can tell by whether it shows up in the devices or shows "FreeFlight".

**Comment by braddward** [comment]
ry835a GPS with Foreflight on iPad and 2 SDR's.  I don't believe that it
was receiving heartbeats when it happened.

Both times I was alone so I couldn't spend much time messing with it.  I
did try the web interface today and that didn't respond.  It seemed like
the entire Pi was locked up.

Except for this I will say that the GPS has been working more reliably.

**Comment by jzeevi** [comment]
No logs, but that happened to my unit today every hour or so....

Josef

**Comment by jzeevi** [comment]
In to different flights, this has happened. FWIW, it may be time related rather than "powered on" related. In both flights, the Stratux disappeared from Foreflight. 

In flight one, after power up and sync to FF, while waiting for engine warmup, my own ship signal disappeared. I checked the devic settings a Freeflight was gone and the web page wouldn't connect. 

In flight two, "no fix" appeared and the device was also missing.

In both cases a power cycle helped. I couldn't restart as the device wasn't connected to the webui. 

Josef

**Comment by Axtel4** [comment]
I had over three hours of flight time yesterday with Stratux build v0.5r3 a99173a with no lock up issues.  Stratux with 2 SDRs, RY835 stayed connected to DroidEFB on my Smart phone and tablet the entire flight time.

**Comment by GrayFlyer** [comment]
I too have been having this problem going back several versions.  I am running dual SDR's with a RY835.  I recently thought I had it narrowed down to a defective Pi, but have been proven wrong.  It's not the tablet, because I have an iFly 720 and a tablet connected and both lose connection at the same time.  The usual lockup occurs around 45 to 60 minutes, and a reboot fixes it, but it sometimes occurs in as little as 15 minutes.  Power good, temps 90 - 105F.

**Comment by cyoung** [comment]
@braddward - could you include /var/log/stratux.log?

**Comment by braddward** [comment]
Hi Chris,

It's in the uploaded file. Stratux.log is in the root and I included the stratux folder as well.

**Comment by braddward** [comment]
But - here's the link to just the stratux.log just in case.
https://app.box.com/s/in6a072weoisv1eqg2jp4pju0gscw9dv

**Comment by Ergonomicmike** [comment]
Just got back home. The stratux hung again on the way home. (Rats - I forgot to enable logging on first boot. It's running a 16GB card. I'm goona leave logging on until v1.0.) I had to pull the SDR and the power again to get it back up. And it came up in 1090, even tho I shut it down in 978.

Since it never used to do this - and I had flown v0.5b3 when it first came out - I'll re-image and re-do the dump1090 change. Unfortunately, we have 5 days of rain forecast this coming week, so no flying/testing  for a while.

I notice others having similar problems all of a sudden. Wouldn't be a Y2016 thing, would it?

**Comment by Ergonomicmike** [comment]
I powered the Stratux back up here at home just in case I had enabled logging after the first flight. I noticed that a lot of the logs wouldn't have done any good because they were zero length. That is, if the SDR doesn't initialize, then no data to log. Is there a diagnostic log that logs the steps the Pi is going thru? Should I set debug true in the config file?

**Comment by skypuppy** [comment]
dmesg?

**Comment by Ergonomicmike** [comment]
@skypuppy I had to look up dmesg. (Will forever be a noob w/ Linux.) You'll have to talk me thru it.

I read that it needs to run syslog in the backgound to capture an output file.  Does syslog currently run on the pi? If so, is the dmesg logfile in the list of other log files?

**Comment by skypuppy** [comment]
from a text console, just enter "dmesg" and if it is logging (and it is
usually turned on) you should see a lot of Greek flowing down your
screen.  If you "dmesg | grep <any stratux term>" it should show you
what happened with stratux at the kernel level.  For example, "dmesg |
grep rtl*" will show you all the kernel messages related to the rtl-sdr
radios.  Or, "dmesg | grep *gdl90" and so on.
Since last boot.

Skypuppy

**Comment by skypuppy** [comment]
And I believe Raspbian has it on by default.  (Raspbian is a version
(currently) of Debian Linux specifically modified for the RPi 1 and 2.  Btw.

**Comment by Ergonomicmike** [comment]
Tnx. But the best I will ever be able to do is to get a log file from my crashed system and post it on the web for you guys to analyze. (Clint Eastwood said that a man has got to know his limitations. That's mine! I'm pushing it posting in the github.)

**Comment by jpoirier** [comment]
On Sat, Jan 2, 2016 at 10:23 PM, Ergonomicmike <notifications@github.com>
wrote:

> Tnx. But the best I will ever be able to do is to get a log file from my
> crashed system and post it on the web for you guys to analyze. (Clint
> Eastwood said that a man has got to know his limitations. That's mine! I'm
> pushing it posting in the github.)
>
You'll be surprised at how quickly you move up the tech continuum...

**Comment by skypuppy** [comment]
I was dragged kicking and screaming all the way up to rung 3 out of
100.  Working on rung 4.

**Comment by skypuppy** [comment]
The more you learn, the more you learn what you don't know.  <sigh> 
It's an exponential growth factor.

I'm just crazy enough to enjoy the ride, though!

**Comment by cyoung** [comment]
@braddward, @GrayFlyer, @Ergonomicmike - I've just put up v0.5b5, give it a try.

**Comment by Ergonomicmike** [comment]
Just saw it. Will write image and head to the airport in the rain.

**Comment by braddward** [comment]
Thanks.  I'll give it a try hopefully this weekend if the weather
cooperates!

**Comment by cyoung** [comment]
Any updates?

**Comment by braddward** [comment]
I flew today for about 1.5 hours with no issues on the latest image.

**Comment by GrayFlyer** [comment]
Everything is ready for a test flight, except the weather.  Looks like some decent weather tomorrow for a flight with 0.5b5 and the latest iFly beta software.

**Comment by GrayFlyer** [comment]
I flew 0.5b5 today for about 50 minutes with no lock-ups, but the AHRS would not connect with either iFly or the web UI.  I tried a reboot of the Stratux, but still no luck.  It seems like connection with AHRS is hit or miss since 0.5b4 or maybe even before.  Pressure altitude worked just fine and the RY835AI was showing down to 0.7M accuracy on the UI at times.  The best I've seen with 0.5b4 was 2M.

**Comment by Nokomis449** [comment]
That's exactly what I've seen since 0.5b3.  Hit or miss AHRS, but pressure altitude works fine, GPS works fine.  0.5b2 AHRS works fine.

**Comment by GrayFlyer** [comment]
I flew 0.5b5 again today and had a lock-up after 16 minutes in flight (27 minutes of connect time,) rebooted and flew for another 40 minutes without another lock-up.  After the lock-up iFly reported Stratux as disconnected with no heartbeat, but a total of 5 heartbeats accumulated which seems very low for 10 minutes or so in flight.  No AHRS on the UI or iFly, but good pressure altitude.

All in all not a problem losing connection as the iFly 720 seamlessly changes over to it's internal GPS when connection with the ADSB GPS is lost, and a reboot always repairs the connection.
![screenshot_2016-01-13-17-04-12](https://cloud.githubusercontent.com/assets/16521434/12313154/fe469d26-ba32-11e5-9bf5-345c2cae4d63.png)

**Comment by skypuppy** [comment]
Do you have cpu temps in your logs?

Also, very strange, but the RPi2 processor is reportedly susceptible to
UV, to the point of destruction.

Another possibility is a buggy compiler.  Or dips in supply power.  Or
bugs in how C code meshes with go code at compile/link time (even though
go is written in C.)  Or some other intermittent hardware failure.  Do
you perhaps have an Element14 RPi2, the ones made in communist China? 
The RY chip(s) are definitely from China, so they may have intermittent
hardware problems that corrupt the system.

Grasping at straws here.

Skypuppy

**Comment by bradanlane** [comment]
On the possibilities of the lockup, my money is still with the recent changes to compile `librtlsdr` into `gen_gdl90`.

RTLSDR isn't guaranteed to be thread-safe and we have introduced a few ways we could be executing in the library from multiple code paths simultaneously.

**Comment by ssokol** [comment]
WRT the lockup theory from @bradanlane... I've talked to a number of
avionics gurus over the past few years about approaches to writing
software. I get the impression that many of them prefer to build systems as
a cluster of independent single-threaded applications rather than relying
on multiple threads within any one application. Threads -- and especially
threads that are managed by the language runtime -- introduce the potential
for subtle bugs that are often very difficult to reproduce and eliminate.

I'm not suggesting that we make a major architectural change at this point,
but it is something to consider for the future.

On Thu, Jan 14, 2016 at 6:31 AM, bradanlane <notifications@github.com>
wrote:

> On the possibilities of the lockup, my money is still with the recent
> changes to compile librtlsdr into gen_gdl90.
>
> RTLSDR isn't guaranteed to be thread-safe and we have introduced a few
> ways we could be executing in the library from multiple code paths
> simultaneously.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cyoung/stratux/issues/180#issuecomment-171632083>.
>



-- 
Steven Sokol
408 Camelot Drive
Liberty, MO 64068

mobile: +1 816-806-8844
fax: +1 816-817-0441

**Comment by bradanlane** [comment]
I've been testing this very architecture. With the current gen_gdl90, we launch a separate process for handling dump1090. I went ahead and moved rtl_sdr|dump978 to also be a separate process. It doesn't appear to be any more/less CPU and its stable, predictable, and log-able. 

I'm not suggesting Stratux change. I just wanted to test the alternate system architecture.

**Comment by GrayFlyer** [comment]
I've been trouble shooting this crash problem on the hardware end for some time now.  I replaced the China produced Pi with the one from Great Britian with no improvement. I've replaced the SD card, the Edimax wifi dongle, the SDR, and tried several power configurations.  Currently I'm supplying 5v from a 12v switching PS located 2" from the GPIO, with parallel connections; solid red led at all times, even when starting the Lycoming. The CPU was running 73F just a minute before crashing (fan cooled in glove box on a cold evening) my normal CPU temp is around 90F. I recently relocated the Edimax away from the power supply just to be sure.  

![1210151231](https://cloud.githubusercontent.com/assets/16521434/12328563/e23bc436-baa9-11e5-91c0-2d535700fd74.jpg)
![1210151232](https://cloud.githubusercontent.com/assets/16521434/12328564/e2403818-baa9-11e5-8e15-d4c8fb185f09.jpg)
![img_7972](https://cloud.githubusercontent.com/assets/16521434/12328587/014ae884-baaa-11e5-903b-1ea5dda6d562.JPG)
![screenshot_2016-01-13-17-02-21](https://cloud.githubusercontent.com/assets/16521434/12328620/265ca1c6-baaa-11e5-83f4-b7d3755d1d1b.png)

**Comment by gitmoregrits** [comment]
For what it's worth, I had very similar problems with the KMASHI battery (the one initially recommended for the Stratux build).  With the Anker (20000 mah?) battery I have zero lock ups and obviously longer up times due to increased capacity.  I know you're running on your ship power but it may make sense to try an Anker if you have one to see if you exhibit any different behavior.  I have flown about 5 hours now with the Anker and haven't had one lock up where as before I was seeing lockups approximately every 30-45 mins.

**Comment by GrayFlyer** [comment]
Thanks for your observation.  I also have a "portable" Stratux that
    runs on a 10ah Anker battery and I don't recall having a lock up
    with it, but I seldom run it for over 20 minutes because of high CPU
    temps on the glare shield.  I wonder what the power supply set up is
    for some of the others that have experienced lock ups is?
    
    I'll give my portable Stratux some more air time and see how it
    fares.
    
    
    
    On 1/14/2016 11:12 AM, gitmoregrits
      wrote:
    
    
      For what it's worth, I had very similar problems with the
        KMASHI battery (the one initially recommended for the Stratux
        build). With the Anker (20000 mah?) battery I have zero lock ups
        and obviously longer up times due to increased capacity. I know
        you're running on your ship power but it may make sense to try
        an Anker if you have one to see if you exhibit any different
        behavior. I have flown about 5 hours now with the Anker and
        haven't had one lock up where as before I was seeing lockups
        approximately every 30-45 mins.
      —
        Reply to this email directly or view
          it on GitHub.

**Comment by skypuppy** [comment]
I'm moving that way toward major structural changes via c/c++ and Qt,
with POSIX threads, signals and slots, and using monolithic software. 
But due to my physical limitations it could be months or years before I
have anything even remotely usable.  Also, NOT using the RY parts and
trying to not use any Chinese parts.

If that is reason to ban me, then so be it.  Meanwhile, in the real
world, I'm still using 0.4r4.

Skypuppy

**Comment by skypuppy** [comment]
Your results seem to confirm cyoungs assertions that it is code
related.  <sigh>

Skypuppy

**Comment by Ergonomicmike** [comment]
@ssokol I'm glad someone much more knowledgeable than I mentioned that single threads are safer than multi. I can't bring anything to the table except to say that I got burned once in a college project by a race condition. Who knows what's going on inside a mult-core chip with stuff being swapped in/out of internal caches.

I think it was ssokol who provided a command line switch for some operation that set how many cores would be used. Perhaps as an easy, first pass way to implement safer threads would be to set any operation that allows for single thread set to single thread? (Are there some processes that require more than one core?)

**Comment by skypuppy** [comment]
maybe it's time to implement resource locks.

**Comment by jzeevi** [comment]
Single threads are safer, but less responsive and, as noted, demand a deliberate and thoughtful approach to the SW design. However, the multi-core comment isn’t applicable to a good design as the HW’s just there to support the effort. Poor design (or no design) is a rat-hole that once headed down, is difficult to get out of. (gah – what horrible grammar). Multi-threaded designs are easy enough to accomplish and don’t require 40 years of embedded experience to understand. Always preferred in complex systems where the processing is less related and more parallel. On a multi-core system, a multi-threaded design wins hands-down except for the most trivial.

 

I think in your note you may be confusing single core with single thread. But, most reasonable OS’s (linux especially) have good handles on process to core assignment although it usually requires the SW to request running on a different core. (I did some ARM A57 development that had 30 threads (two processes) on 16 cores with no trouble doing real-time network traffic at gigabit phy rates). 

 

I’m fond of saying (not sure where I got it): If you don’t know where you’re going, any road will get you there.

 

Note: I suspect SW engineering isn’t the problem here, but implementation “oopses”. J

**Comment by Ergonomicmike** [comment]
You're correct - I confused threads with cores.

**Comment by Ergonomicmike** [comment]
And IIRC, early Intel uP's ran faster with their hyper-threading/predictive processing off rather than on. Maybe those days are behind us. But I'll be forever leery of race conditions. And since this issue is about "random" hangs, that's my usual suspect.

**Comment by skypuppy** [comment]
compilers are much better nowadays.  :)  We need a good debugging
session where we can trap  when two or more sdr messages arrive during
the same function processing phase.  It could be that one sdr message is
clobbering the current cycle.

Impossible to replicate this problem with the current information, but
it must be done to prevent the app from crashing.

**Comment by jzeevi** [comment]
Yes, those days are well behind us. Ah, the good old days when Intel couldn’t do floating point (our team discovered that in parallel with several others).

 

It is wise to be leery of race conditions! But that’s not a reason to avoid multi-anything programming. One only has to be thoughtful and deliberate about it.

 

And, since we’ve drifted so far off topic, I call Pizza on myself. J

**Comment by cyoung** [comment]
@GrayFlyer - looking at your screenshots, I see you're using an Amazon Fire. Have not tested on this platform yet. Do you have another device (Android or iOS) that you're able to connect to stratux to see if you experience the same?

**Comment by GrayFlyer** [comment]
Hey Chris,
    
    Yes, I currently run iFly and Avare on my Samsung Tab S and have ran
    iFly on my iPad 2, all with the same experience.  The Amazon Fire
    that I'm running iFly on is a rooted $49.99 model, lots of bang for
    the buck!  I run the tablets in parallel with my iFly 720, and also
    use them for access to your UI.  When the problem occurs it affects
    both connected devices.
    
    FWIW, I just flew with my portable battery powered Stratux for about
    an hour with no problems on 0.5b5 with my iFly 720 and the Fire for
    the UI.
    
    Thanks for looking into this!
    
    Dave
    
    On 1/14/2016 5:39 PM, cyoung wrote:
    
    
      @GrayFlyer - looking at your
        screenshots, I see you're using an Amazon Fire. Have not tested
        on this platform yet. Do you have another device (Android or
        iOS) that you're able to connect to stratux to see if you
        experience the same?
      —
        Reply to this email directly or view
          it on GitHub.

**Comment by cyoung** [comment]
Ok, so here's what we have currently on the lockups -
1. @braddward was using FF on v0.5b3, upgrading to v0.5b5 appears to have fixed it and the problem has not resurfaced.
2. @GrayFlyer, @Ergonomicmike are using iFly on various platforms and the problem is still persisting.
3. @Axtel4 reports v0.5b3 did not have any issues with DroidEFB.
4. @jzeevi reported issues with an stratux unknown version with FF.

So @jzeevi, @GrayFlyer, and @Ergonomicmike remain now.

@jzeevi - have you tried v0.5b5 with FF? What version was giving you issues originally?

@GrayFlyer, @Ergonomicmike - I'm going to branch off your reports and consider that this may actually be a separate issue with the way stratux works with iFly.

Going to go flying today with iFly GPS 9.4.36b on Android 4.2 (cheap tablet) and see what happens - first time trying out the app.

**Comment by jzeevi** [comment]
Initially this was with .5b3. Over the weekend I flew with .5b5 and had a similar thing. I wound up rebooting to tees tableau the connect. 

Josef

**Comment by GrayFlyer** [comment]
@cyoung.  Be sure to turn on ADSB logging (under wireless devices)  After your flight look in the iFly files under "user" for it and also #logs (on by default)  Even if you don't have a problem, you may find some clues in those files.

**Comment by cyoung** [comment]
I've added http://192.168.10.1/getClients to the current master branch for debugging - suspect it has to do with sleepmode.

**Comment by Ergonomicmike** [comment]
Sleep mode in Stratux? 'Cause my tablets don't sleep in flight.

Hope you played with iFly on the ground a bunch. Lots of ways to customize it to your liking, which can be confusing at first. Once configured as you like it, it flows very nicely in flight.

**Comment by cyoung** [comment]
Sleep mode in stratux, described here: https://github.com/cyoung/stratux/blob/master/notes/app-vendor-integration.md

**Comment by Ergonomicmike** [comment]
Thx. I'll check it out. Meantime, doing some testing on the couch. Running Plain Vanilla Single Dongle Stratux, v0.5b5. My Asus Lollipop tab with Avare. My Nook HD+ running custom ROM CM 10.2.1 (JB) with iFly. iFly continues to show traffic even when I disable WiFi on the Nook!

**Comment by Ergonomicmike** [comment]
Cannot get iFly to connect to Stratux when Avare's I/O plug-in is listening. So is the problem with Avare or iFly?

Been running three tablets now on the couch. Two on iFly, one on Avare. No loss of Wi-Fi connectivity so far.  How do I force the Wi-Fi to sleep to see what happens after waking up?

**Comment by Ergonomicmike** [comment]
Cannot get iFly to connect to Stratux when Avare's I/O plug-in is listening. So is the problem with Avare or iFly?

Been running three tablets now on the couch. Two on iFly, one on Avare. No loss of Wi-Fi connectivity so far.  How do I force the Wi-Fi to sleep to see what happens after waking up?

**Comment by Ergonomicmike** [comment]
Cannot get iFly to connect to Stratux when Avare's I/O plug-in is listening. So is the problem with Avare or iFly?

Been running three tablets now on the couch. Two on iFly, one on Avare. No loss of Wi-Fi connectivity so far.  How do I force the Wi-Fi to sleep to see what happens after waking up?

**Comment by Ergonomicmike** [comment]
Cannot get iFly to connect to Stratux when Avare's I/O plug-in is listening. So is the problem with Avare or iFly?

Been running three tablets now on the couch. Two on iFly, one on Avare. No loss of Wi-Fi connectivity so far.  How do I force the Wi-Fi to sleep to see what happens after waking up?

**Comment by Ergonomicmike** [comment]
Cannot get iFly to connect to Stratux when Avare's I/O plug-in is listening. So is the problem with Avare or iFly?

Been running three tablets now on the couch. Two on iFly, one on Avare. No loss of Wi-Fi connectivity so far.  How do I force the Wi-Fi to sleep to see what happens after waking up?

**Comment by Ergonomicmike** [comment]
Cannot get iFly to connect to Stratux when Avare's I/O plug-in is listening. So is the problem with Avare or iFly?

Been running three tablets now on the couch. Two on iFly, one on Avare. No loss of Wi-Fi connectivity so far.  How do I force the Wi-Fi to sleep to see what happens after waking up?

**Comment by Ergonomicmike** [comment]
Cannot get iFly to connect to Stratux when Avare's I/O plug-in is listening. So is the problem with Avare or iFly?

Been running three tablets now on the couch. Two on iFly, one on Avare. No loss of Wi-Fi connectivity so far.  How do I force the Wi-Fi to sleep to see what happens after waking up?

**Comment by Axtel4** [comment]
> Cannot get iFly to connect to Stratux when Avare's I/O plug-in is listening. So is the problem with Avare or iFly?

On the Android side, I have not been able to get Avare or Droid EFB to connect on the same device at the same time since 0.4r2.  I have to stop whichever App was started first in order to get the other App to connect.

> How do I force the Wi-Fi to sleep to see what happens after waking up?

I am only familiar with Android, but what happens if "Keep Wi-Fi on when screen is off" is set to "No".  Would that cause Stratux to "sleep" when the devices disconnect W-Fi at that time?

**Comment by cyoung** [comment]
Was able to reproduce this bug, completely locking up wifi. It seems like a hostapd problem, looking into it.

**Comment by Ergonomicmike** [comment]
Okay, I put my tablet to sleep for up to 10 minutes. (I made sure that Wi-Fi keep alive was set to No.) Perhaps I don't understand what the "Sleep mode" is supposed to do to the Stratux. The blue light flashes as much as usual. It occurred to me that one difference here at home vs. the plane is that there are routers around here. Is the Stratux pinging those and getting a response that makes it think a client is awake? In any event, I even turned the WiFi off on the tablet, but could not reproduce the Wi-Fi disconnect problem that I had in the plane. Looking forward to v0.5v6.

**Comment by Axtel4** [comment]
I have been trying to duplicate the "lock up" issue, but I have not been successful.  Before testing I did a git pull on the Commit that had the /getClients changes and recompiled the code.  I started Stratux and the EFB App on my tablet.  I then closed the tablet to "put it to sleep."  ( It is set for Keep Wi-Fi on during Sleep / Never.)  This morning I opened up the tablet and it connected to Stratux and started displaying traffic information on the App.  I may not be understanding how the Sleep Mode works.

In doing the above, it also lead me to a question about the /getClients feature.  Last night and this morning I saw there were 4 clients connected to Stratux, however I could only find one client actually connected, the tablet.  This evening I shut off all W-Fi devices except the tablet, cleared the browser cache, cold Booted Sratux, and reloaded the /getClients page.  I was surprised to see the /getClients page still showing the came 4 clients connected.  So, my questions,

1a) When does Stratux flush the /getClients data so it only shows what is actually connected?

1b) Is there a file stored somewhere on Statux with the previous clients that should be deleted before quarrying for refreshed data?

2) The date / Time stamp shows "0001-01-01T00:00:00Z" but the R-Pi clock and the GPS data shows the correct time stamp.  Is there an error in my R=Pi?

3) When refreshing the /getClients page, the "Sleep Flag:" for the active client IP address toggles between TRUE and FALSE with subsequent refreshes.  However, it doesn't toggle with each refresh.  Shouldn't this persist at one state (i.e. FALSE) when the client is actively connected?

This is a copy of a /getClients load:
(note: I am using Port 43211 for DroidEFB)
>{"192.168.10.11:43211":{"Conn":{},"Ip":"192.168.10.11","Port":43211,"Capability":5,"LastUnreachable":"0001-01-01T00:07:52Z","SleepFlag":false},"192.168.10.11:49002":{"Conn":{},"Ip":"192.168.10.11","Port":49002,"Capability":2,"LastUnreachable":"0001-01-01T00:07:53Z","SleepFlag":true},"192.168.10.12:43211":{"Conn":{},"Ip":"192.168.10.12","Port":43211,"Capability":5,"LastUnreachable":"0001-01-01T00:00:00Z","SleepFlag":true},"192.168.10.12:49002":{"Conn":{},"Ip":"192.168.10.12","Port":49002,"Capability":2,"LastUnreachable":"0001-01-01T00:00:00Z","SleepFlag":true},"192.168.10.13:43211":{"Conn":{},"Ip":"192.168.10.13","Port":43211,"Capability":5,"LastUnreachable":"0001-01-01T00:00:00Z","SleepFlag":true},"192.168.10.13:49002":{"Conn":{},"Ip":"192.168.10.13","Port":49002,"Capability":2,"LastUnreachable":"0001-01-01T00:00:00Z","SleepFlag":true},"192.168.10.14:43211":{"Conn":{},"Ip":"192.168.10.14","Port":43211,"Capability":5,"LastUnreachable":"0001-01-01T00:00:00Z","SleepFlag":true},"192.168.10.14:49002":{"Conn":{},"Ip":"192.168.10.14","Port":49002,"Capability":2,"LastUnreachable":"0001-01-01T00:00:00Z","SleepFlag":true}}

Thanks.

**Comment by Ergonomicmike** [comment]
I have noticed that the number of clients in the web UI shows the max number of different clients there have been. (As opposed to a count of current clients.) So, if you are showing "4," have you had 4 different tablets/phones talking to your Stratux since last image?

**Comment by kdknigga** [comment]
Stratux knows its clients by DHCP leases.  When a client connects to the stratux wifi dhcpd gives the client an IP address and records this in /var/lib/dhcp/dhcpd.leases.  Stratux reads that file to know who its clients are.

From `man 8 dhcpd`:

> New  leases  are  appended to the end of the dhcpd.leases file.  In order to prevent the file from becoming arbitrarily large, from time to time dhcpd creates a new dhcpd.leases file from its in-core lease database.  Once this file has been written to disk, the old file is renamed dhcpd.leases~, and  the  new  file  is  renamed dhcpd.leases.   If the system crashes in the middle of this process, whichever dhcpd.leases file remains will contain all the lease information, so there is no need for a special crash recovery process.

Unfortunately, I can't find how often this refreshing process happens.  It's rare enough that old clients do tend to pile up.

**Comment by Axtel4** [comment]
Thanks for pointing me in the right direction. Also the dchp document answered the other part of my question:

> In order to keep track of leases across system reboots and server restarts, dhcpd keeps a list of leases it has assigned in the dhcpd.leases(5)file. Before dhcpd grants a lease to a host, it records the lease in this file and makes sure that the contents of the file are flushed to disk. This ensures that even in the event of a system crash, dhcpd will not forget about a lease that it has assigned.

**Comment by Ergonomicmike** [comment]
So a stupid question: Why does the Stratux need to do this? If there's a crash (or even just a power down restart), won't the Stratux just reconnect like my router does?

**Comment by kdknigga** [comment]
I don't understand your question.  Why does Stratux need to know who its clients are?  Why does dhcpd need to keep track of leases?

**Comment by Ergonomicmike** [comment]
Yes. Say it doesn't keep track of clients & leases thru power down resets. Isn't that the same as starting with a fresh new image? Or is the implication that the clients would need to be reset as well to reconnect?

**Comment by kdknigga** [comment]
I don't *think* it would be a problem if dhcpd lost track of the leases on reboot.

**Comment by Ergonomicmike** [comment]
On my home routers, I dedicate certain IP addresses to clients that have static addresses or that I need to be on repeatable addresses. But I don't even know (or care) what the IP's are on my Stratux.

**Comment by kdknigga** [comment]
dhcpd keeps track so it doesn't give out duplicate IPs, not so that devices consistently get the same IP.

**Comment by Ergonomicmike** [comment]
Okay. But after a power loss, all bets are off and the IP assignments start anew.

**Comment by Nokomis449** [comment]
Not exactly - re-read what kdknigga posted above, especially the the last sentence:

New leases are appended to the end of the dhcpd.leases file. In order to prevent the file from becoming arbitrarily large, from time to time dhcpd creates a new dhcpd.leases file from its in-core lease database. Once this file has been written to disk, the old file is renamed dhcpd.leases~, and the new file is renamed dhcpd.leases. If the system crashes in the middle of this process, whichever dhcpd.leases file remains will contain all the lease information, so there is no need for a special crash recovery process.

**Comment by Ergonomicmike** [comment]
Right. I understand. My stupid question was why the file has to be written in the first place? Let the Stratux simply start up as it does with a fresh image. I don't know if there's any advantage to not writing the file. Just questioning everything. Probably time for Chris to weigh in.

**Comment by cyoung** [comment]
The IP assignments are not relevant, there is no good reason for keeping the DHCP leases across reboots. It was more or less a convenience thing. Stratux watches the leases file immediately on startup. The file was indexed previously by its inode and the DHCP startup process creating/removing the file was causing some problems with that. So the lease time was increased. We moved away from watching the file by its inode but didn't roll back the DHCP lease time. That's something we could do to increase resource efficiency.

**Comment by cyoung** [comment]
>1a) When does Stratux flush the /getClients data so it only shows what is actually connected?

Long time. Really only want to watch sleep true/false here for now.

>
>1b) Is there a file stored somewhere on Statux with the previous clients that should be deleted before quarrying for refreshed data?
>
>2) The date / Time stamp shows "0001-01-01T00:00:00Z" but the R-Pi clock and the GPS data shows the correct time stamp. Is there an error in my R=Pi?
>

The clock used for this starts at 0 and counts the seconds since stratux started, it's not supposed to be real time. Your RPi is fine.

>3) When refreshing the /getClients page, the "Sleep Flag:" for the active client IP address toggles between TRUE and FALSE with subsequent refreshes. However, it doesn't toggle with each refresh. Shouldn't this persist at one state (i.e. FALSE) when the client is actively connected?

This is most interesting. What operating system was the client running, was an EFB app active on the screen, screen turned on?

**Comment by cyoung** [comment]
Factors that increase your chances of reproducing this bug: connecting to the WiFi at a slower speed (put the tablet some ways away and check the connection speed), increasing network output (-replay -uatlog /var/log/stratux/XXXX-uat.log.gz -speed 10).

**Comment by skypuppy** [comment]
When Linux issues dhcp leases, it acts as a server.  This is not a
stratux thing, but stratux uses that data.

**Comment by Ergonomicmike** [comment]
Yes. But if it's not necessary, then one less thing to write if Chris goes RO.

**Comment by Axtel4** [comment]
> What operating system was the client running

Android KitKat (Table) or Android Lollipop (Smart Phone)

>  EFB app active on the screen, screen turned on?

Yes and Yes.  

I had the App (DroidEFB) running on the tablet and had sshed into Stratux via a wired connection on my desktop.  I launched a browser session the desktop to monitor /getClients.  

Update: I did some more testing when I got home this evening and maybe this will make sense to you.  With the App launched and "in focus" I kept refreshing the browser.  What I saw as the 192.168.10.11 Port 43211 SleepFlag stayed "false" whereas the 192.168.10.11 Port 49002 SleepFlag would intermittently toggle between "false" and "true."  As long as the App is running, the Pot 43211 SleepFlag remained "false."  It didn't matter if it is in focus or is minimized.  When I disconnected the tablet from the Stratux Wi-Fi or closed (exit) the App, the SleepFlag would toggle "true."  (DroidEFB dosn;t appear to support the Stratx AHRS at this time.)

**Comment by Axtel4** [comment]
> Factors that increase your chances of reproducing this bug: connecting to the WiFi at a slower speed (put the tablet some ways away and check the connection speed), increasing network output (-replay -uatlog /var/log/stratux/XXXX-uat.log.gz -speed 10).

Ok, I got the first half, but I am not ashamed to admit the second half went over my head.  (You guys have been stretching my knowledge with this project.)  I think the last half invokes the replay of previously recorded data.  If so, is only this entered on the command line? 

> -replay -uatlog /var/log/stratux/XXXX-uat.log.gz -speed 10

Also, I have been recording uncompressed data.  Can I just replay the uncompressed log, or do I have to compress it before replaying?

Thanks.

**Comment by Ergonomicmike** [comment]
I think that Chris wants you to stress the system by increasing the speed of logging so that the Wi-Fi sleep true/false flag vacillation might be reproduced at will.

**Comment by Ergonomicmike** [comment]
I flew with iFly beta v.9.4.40b today and Stratux v0.5b5 with Chris's CPU & network tweaks (per https://github.com/cyoung/stratux/commit/5aca8adea2b4f8c6fe7baf312088bafa152a73be). I had to pull power on the Stratux a few times  - no heartbeat was the report from iFly's screen. But I was so busy doing instrument approaches, I wasn't really able to give the Stratux the attention I normally do. So I'm going to call this "inconclusive" for now.

**Comment by Axtel4** [comment]
I think I saw the same thing today while testing and I too had the latest Wi-Fi changes. There were a few times Stratux stopped sending data to two phones and a tablet. The Web UI stopped updating and the EFB App froze and the SSH session would drop.  A couple of times I power cycled Stratux and recovered.  The third time I just cycled the Wi-Fi off to on on the tablet  while leaving Stratux running and the Wi-Fi connection restarted on tablet and phones. The interesting thing is I didn't do anything on the phones, it just started receiving Wi-Fi after cycling Wi-Fi on the tablet. After two more times of cycling the tablet Stratux stayed connected for over 2 hours until I shut it down.

**Comment by GrayFlyer** [comment]
@Ergonomicmike   Was your iFly the only device connected to the Stratux WiFi?

**Comment by skypuppy** [comment]
If anyone is near Huntsville, Alabama and wants someone to monitor their
unit while flying, I'd be happy to lend a hand. 

Skypuppy

**Comment by Ergonomicmike** [comment]
@GrayFlyer  No, I had two Android iFly 40b tablets listening during the flight. I wonder if two iFly tablets running the same version, licensed to the same user confuses the Stratux?

@skypuppy  I'll keep that in mind. What is there to see while in Huntsville?

@cyoung If I hypothesize that this is a DHCP lease problem (perhaps because the Pi is running on its own time zone?), is there a way that I can set static IP addresses in the Stratux? I would then set the wireless connections statically on my tablets to see if the connections stay connected.

**Comment by Nokomis449** [comment]
@skypuppy Which airport?  I'm based at 8A0 and work in Huntsville.

**Comment by kdknigga** [comment]
> I'll keep that in mind. What is there to see while in Huntsville?

Uhhh... The US Space and Rocket Center, home of Space Camp?

**Comment by kdknigga** [comment]
Also known as my happiest memory from junior high.

**Comment by skypuppy** [comment]
Mike, Space and Rocket Center is a movie/museum/rides and is not
expensive.  We have not one but TWO Saturn 5 rockets, one upright, one
on it's side (and another at I-65 rest area near Tennessee.)  We have
the first wooden shuttle mockup, too.  Some rides.  Space Camp is
located on their campus as well.  Oh, the theater is IMAX.
also, HSV has museums, couple small art galleries, girlie bars, regular
bars :) , and the usual for a city of about 150,000.  Sometimes NASA's
Marshall Space Flight Center gives tours through their portion of
Redstone Arsenal.
The civic center has top shelf shows and conventions, too.  Check on Von
Braun Center online. 
There is an astronomy club with open shows now and then.
And some fantastic motorcycle twisties nearby, too.  :)

DHCP leases are persistent as long as they are connected.  Sometimes for
years.

Skypuppy.

**Comment by skypuppy** [comment]
I fly mostly out of 3M5, but for stratux tinkering, I'll come to
whichever you need me.  Where is 8A0?  Is that Decatur?  Cullman?  I
forget.  :(

**Comment by Axtel4** [comment]
This is an update to my post yesterday.   I have been able to replicate the "lockup" issue today at home but it seems to be intermittent.

>Set up:
>ASUS TF-300 Tablet (Kitkat) or LG G3 (Lollipop)
> DroidEFB and the Web UI running
 >Stratux 0.5b5 w/ through commit f5bed44 including a manual update to include commit 5aca8ad.

The tablet was connected via Wi-Fi to Stratux with DroidEFB and the Web UI page running.  The status page of the Wed UI is monitored for activity (i.e. Uptime advancing).  Three times today the Stratux dropped connection to the tablet with the Web UI Uptime frozen at the time connection was lost.  Also the EFB App dropped all ADS-B data while the Wi-Fi page on the tablet shows that Stratux is still connected.

To recover, all I have done is cycle the Wi-Fi off to on on the tablet while leaving Stratux running.  (I have also done the same thing with my smart phone when it was the connected client.)  The client reconnects to Stratx and the Web UI and the EFB App will again display data.

The interesting this I have found is that if I have a parallel Ethernet connection to my desktop through my router, Stratux has never dropped the W-Fi connection.  Also note, the Tablet is set to never sleep.

**Comment by Ergonomicmike** [comment]
I finally managed to get the build process to work and am now running v.0.5b5 with 28 commits. (More on the build process over in the https://github.com/cyoung/stratux/issues/216 thread.)

@Axtel4, do you get the lockups on 1090, 978, both? I ran 1090 for hours today with two tablets running iFly Android and it did not lock up. I have anecdotally noticed that it seems like I only get lockups in flight when monitoring 978. (BTW, when I tried 978 today, I saw my first ever 978 traffic from the ground, without towers!) I don't know why 978 traffic should make a difference when it locks up - more traffic than 1090?

Also @Axtel4 (and @cyoung), there's a new getClients commit for debugging (https://github.com/cyoung/stratux/issues/216.) How do we use that to help Chris? (I looked for a log called Clients, but didn't see it.)

While running my tests today, I noticed that one of my tablets shows when the Wi-Fi is transmitting and receiving. It was transmitting while hooked up to Stratux. 

I'll show my ignorance here: I thought that the Stratux was sending UDP packets, which don't need an ACK. So why is my tablet talking to the Stratux? Responding to a ping?

On the assumption that it is supposed to check in every now and then, could using the same software (iFly) on two different tablets cause two identical transmissions at the same time, confusing the Stratux? Anecdotally, it seems like I only get this lock up problem in flight when I run more than one tablet using iFly. (Although Chris says he got a lock up in his test. Presumably only one client?)

@skypuppy Yeah, I know that DHCP leases are supposed to persist. But since we're having a problem that seems to be network related, I thought that bypassing the DHCP client with static addresses might help eliminate (or isolate) DHCP as the problem.

**Comment by GrayFlyer** [comment]
Just looking back at some screenshots during a loss of heartbeat and GPS position event and noticed that after iFly reported the losses, that the traffic targets keep on moving for 40 seconds or more.  I don't think the  GPS itself lost lock, as the lock LED kept blinking and it had excellent stats before and after.
Is there a clue here?

![screenshot_1](https://cloud.githubusercontent.com/assets/16521434/12570852/67f0ddac-c3a9-11e5-8266-308ed24b328c.jpg)
![screenshot_2](https://cloud.githubusercontent.com/assets/16521434/12570853/67fc851c-c3a9-11e5-86a5-18b2812d4446.jpg)
![screenshot_3](https://cloud.githubusercontent.com/assets/16521434/12570854/6806ef84-c3a9-11e5-946d-1a5ba2bafcae.jpg)
![screenshot_4](https://cloud.githubusercontent.com/assets/16521434/12570855/680a70e6-c3a9-11e5-97e1-48c0a486514b.jpg)

**Comment by Axtel4** [comment]
I am not sure how iFly works, but it may be "coasting" the traffic targets after they have been lost from the source.

**Comment by GrayFlyer** [comment]
@Axtel4   I think you are right, I can see the targets sliding down the "speed vector" speed, and notice that even though they are indicating a decent or climb, the altitude delta never changes.  I've asked iFly anyway though.

**Comment by ssokol** [comment]
That's uncanny. I also have a Huntsville connection - my day job is with
Digium, the makers of the Asterisk open source telephony engine. I've flown
in and out of Moontown once. I've been into KMDQ many times. Nice part of
the country.

Any chance you know a Huntsville-area pilot named Mark Spencer? He's
Digium's founder and now the CEO at Avilution - the people making XFS.

On Sun, Jan 24, 2016 at 2:23 PM, skypuppy <notifications@github.com> wrote:

> I fly mostly out of 3M5, but for stratux tinkering, I'll come to
> whichever you need me. Where is 8A0? Is that Decatur? Cullman? I
> forget. :(
>
>
> On 01/23/2016 10:41 PM, Nokomis449 wrote:
> >
> > @skypuppy <https://github.com/skypuppy> Which airport? I'm based at
> > 8A0 and work in Huntsville.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/cyoung/stratux/issues/180#issuecomment-174251850>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cyoung/stratux/issues/180#issuecomment-174332703>.
>



-- 
Steven Sokol
408 Camelot Drive
Liberty, MO 64068

mobile: +1 816-806-8844
fax: +1 816-817-0441

**Comment by skypuppy** [comment]
Sorry, the name doesn't ring a bell.
And where is 8A0?

**Comment by Nokomis449** [comment]
8A0 = Albertville Airport, on Sand Mountain just south of Guntersville.  As for things to do around here, add these to the list:  Lake Guntersville is arguably the fishing capital of the south and a boating playground; the Jack Daniels Distillery is fairly close in Lynchburg, TN, Campbell Field is a short flight to a grass strip and great food, the white sand beaches of Gulf Shores, AL are a nice flight away, and as mentioned before, some great motorcycle roads.

**Comment by cyoung** [comment]
https://github.com/cyoung/stratux/releases/tag/v0.5b6

**Comment by Ergonomicmike** [comment]
Okay, I flew v0.5b6 today. 2 hours going, 2 hours coming back.

I think I have good news on the connectivity issue - on my end at least. Although I had to power down reset the Stratux a couple times on the way out, I think I've got the problem isolated and I found a way to keep the problem from happening. And it seems to be unique to my setup.

There is something goofy with my iFly Nook HD+, which uses a custom ROM. When it is connected to the Stratux via Wi-Fi AND when it it is getting GPS data from my other iFly tablet via BlueTooth, it will reliably (repeatably) lock up the Stratux network (to the point that the Stratux UI won't work) about every 15 minutes. 

The Developer of the ROM acknowledges that you sometimes have to initialize Wi-Fi before BT in the ROM to get things to work. Also, in my firewall on the Nook, I have to enable LAN ports to get Stratux to work. (As opposed to just the Wi-Fi ports.) So there's something goofy in that ROM about how the Developer wired internal networks. 

The Stratux connection would get totally locked out from the Nook, to the point where "Wi-Fi Analyzer" wouldn't even see the signal. (That is, scanning for the Stratux was a no show.)  Although Wi-Fi Analyzer saw the signal on my other tablet, the Stratux wasn't able to talk to it during this time.) Disabling BT on the Nook and then restarting WiFi on the Nook would fix the problem. (And then the other tablet would come back on line by itself.) And then the problem would repeat a little while later.

So on the way home, I used my Tom-Tom GPS dongle to BT GPS data to the Nook HD+ instead of my other tablet. I had both tablets connected to the Stratux, and I didn't lose the connection for the entire flight. 

That's only a sample of one flight. But the result is consistent with my testing on the couch, where I could run both tablets for hours connected to the Stratux without problems. (I had enabled GPS on the one tablet, but I didn't think to BT it over to the Nook during my couch testing.) 

For now I think this issue is fixed for me. I'll report back if it comes back.

**Comment by cyoung** [comment]
Excellent, thanks.

@GrayFlyer and @jzeevi remain now, will you guys be able to try 0.5b6?

**Comment by jzeevi** [comment]
Later today....

**Comment by braddward** [comment]
I'm flying on Saturday and will give it a try as well.

**Comment by Ergonomicmike** [comment]
BTW, on the flight back and forth to Death Valley, I was switching back and forth between 1090 and 978. That worked well too. I did not have to reboot to get the dongle to change. (I really need to buy another SDR. There's a lot to of people flying around our altitudes with 1090 to make it worth the $20.)

**Comment by cyoung** [comment]
Don't forget a good cable and battery if you're going dual SDR.

**Comment by Ergonomicmike** [comment]
Thanks. I found a shorter USB to micro cable at Fry's which helps keep the red light solid longer for my current setup. But the LED still goes out when the KMASHI is at 2 bars or less. I plan to buy the VK-172 also, so I will probably have to buy the Anker too. (I was hoping that Anker would address that voltage dip issue when we shutdown the Pi in a Gen 3 before buying.)

**Comment by skypuppy** [comment]
Bizarre

**Comment by GrayFlyer** [comment]
It will be at least a week before I can test 0.5b6 in flight as my
    prop is in the shop for a reseal.
    
    
    
    On 1/28/2016 8:24 AM, Josef Zeevi
      wrote:
    
    Later today....
      
      Sent from my iPad
      
      > On Jan 28, 2016, at 6:22 AM, cyoung
      <notifications@github.com> wrote:
      > 
      > Excellent, thanks.
      > 
      > @GrayFlyer and @jzeevi remain now, will you guys be able to
      try 0.5b6?
      > 
      > —
      > Reply to this email directly or view it on GitHub.
      > 
      —
        Reply to this email directly or view
          it on GitHub.

**Comment by Nokomis449** [comment]
@Ergonomicmike - I've had great luck with a 10 or 12aH Zackery battery from Amazon Prime.  It will charge while discharging, and on the ground it will run 2 dongles, wireless and a VK for over 4 hours before I unplug it And there's still battery life left.  And when I shut it down, it STAYS shutdown. (But wait, that's not all!  Order in the next 30 minutes and you'll also get a Ronco Vegi-Matic absolutely free!)

**Comment by ssokol** [comment]
I'll be doing some testing this weekend.

On Thu, Jan 28, 2016 at 9:20 AM, GrayFlyer <notifications@github.com> wrote:

>
>
>
>
>
> It will be at least a week before I can test 0.5b6 in flight as my
> prop is in the shop for a reseal.
>
>
>
> On 1/28/2016 8:24 AM, Josef Zeevi
> wrote:
>
> Later today....
>
> Sent from my iPad
>
> > On Jan 28, 2016, at 6:22 AM, cyoung
> <notifications@github.com> wrote:
> >
> > Excellent, thanks.
> >
> > @GrayFlyer and @jzeevi remain now, will you guys be able to
> try 0.5b6?
> >
> > —
> > Reply to this email directly or view it on GitHub.
> >
> —
> Reply to this email directly or view
> it on GitHub.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cyoung/stratux/issues/180#issuecomment-176204552>.
>



-- 
Steven Sokol
408 Camelot Drive
Liberty, MO 64068

mobile: +1 816-806-8844
fax: +1 816-817-0441

**Comment by jzeevi** [comment]
Can we get more details on that battery? I wasn’t able to find it in amazon….

**Comment by Nokomis449** [comment]
Sorry, I either mispelled it or was a victim of autocorrect.  Try Jackery with a J.  I have the Jackery Giant and the Jackery Bar.

**Comment by ssokol** [comment]
Let us know how it does. The customer reviews on Jackery's own site are a
bit negative, but if the thing can charge and power things at the same
time, that's a big positive.

On Thu, Jan 28, 2016 at 11:19 AM, Nokomis449 <notifications@github.com>
wrote:

> Sorry, I either mispelled it or was a victim of autocorrect. Try Jackery
> with a J. I have the Jackery Giant and the Jackery Bar.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cyoung/stratux/issues/180#issuecomment-176259333>.
>



-- 
Steven Sokol
408 Camelot Drive
Liberty, MO 64068

mobile: +1 816-806-8844
fax: +1 816-817-0441

**Comment by Nokomis449** [comment]
I've had both Jackery's for about 2 months now and no problems.  I use the smaller one (6aH?)  with a single dongle/GPS Stratux and I just built a dual/GPS Stratux for a friend and used the bigger Jackery for burn in and other testing.  Both  batteries are used for charging phones, tablets, etc. when they're not connected to a Stratux. They're slow to charge, but that could easily be a function of my charger or cable.  I also have a 2-3 year old 10aH Anker that can charge while discharging and it too is a solid performer.  For every Stratux I build for a friend, it seems like two other friends want one.  Since power supply is so critical, I'm thinking about providing a Jackery and cable with each build so if anyone has had a firsthand bad experience with them I'd like to know.

**Comment by Ergonomicmike** [comment]
The following seems counter intuitive to me. I comment for what it's worth.

I use a firewall on my Android tablets. Normally I allow applications to have "Wi-Fi" Permissions. Which is what I expect for a Wi-Fi connection. But with the Stratux on Wi-Fi, I have to allow applications LAN Permission instead. http://postimg.org/image/vsropf9uh/

(If you use a firewall, be advised that you also have to allow the Linux Kernel to have LAN Permission. (Not shown in the screen shot.))

**Comment by Nokomis449** [comment]
LAN = Local.  Perhaps the Stratux is considered a local device, since access to it is over the LAN and not a WAN (internet) connection.  After all, it's on the same subnet (192.168.10.0) as the tablet.

**Comment by cyoung** [comment]
>There is something goofy with my iFly Nook HD+, which uses a custom ROM. When it is connected to the Stratux via Wi-Fi AND when it it is getting GPS data from my other iFly tablet via BlueTooth, it will reliably (repeatably) lock up the Stratux network (to the point that the Stratux UI won't work) about every 15 minutes.
>
>The Developer of the ROM acknowledges that you sometimes have to initialize Wi-Fi before BT in the ROM to get things to work. So there's something goofy in that ROM about how the Developer wired internal networks. 

@Ergonomicmike 

Could it be possible that the Nook and the iFly tablet were using the WiFi network to communicate? Were they both connected to WiFi? There is a known issue with the WiFi setup and high traffic loads - same symptoms.

**Comment by Ergonomicmike** [comment]
Yes, they were on the same network, in the sense of 192.168.10.x. Here's a photo of getClients from my smart phone in flight. http://postimg.org/image/iw547rvzr/c0c75d58/ (Two tablets and my smart phone connected. A third tablet, that I was using for testing at home, is not connected.)

To be clear, I have two Android tablets running the iFly app. (Distinct from iFly's dedicated box, since I've been calling one of them an iFly tablet.) One is the Nook HD+ running JB with a custom ROM and kernel. (With known BT problems.) The other is an Asus ZenPad S 8, running stock LP. (Except rooted and debloated.)

**Comment by Nokomis449** [comment]
Don't know if Mike was doing it, but iFly has the ability to share flight plans between other iFly apps on the network, including the option to push changes to the plan that may be added on-the-fly.  I haven't tried it yet with Stratux but it works great with my SkyRadar.  However, I can't see it requiring a whole lot of WiFi time.  I'll test it out later, but I haven't seen any WiFi drops that started this thread that weren't power source/cable related.

**Comment by Ergonomicmike** [comment]
I know about that feature in iFly, but haven't tried it yet. So, AFAIK, clean Wi-Fi environment. (Although a fiberglass plane w/ no shielding from the xpndr antenna.)

**Comment by Ergonomicmike** [comment]
I got to thinking about BT & Wi-Fi running at the same time. Apparently they used to interfere w/ ea other in the old days. http://www.hp.com/rnd/library/pdf/WiFi_Bluetooth_coexistance.pdf  (circa 2002.) I read that newer modulation techniques are supposed to mitigate this nowadays. Still, IIRC someone here suggested that Stratux allow the ability to change channels. Might be a worthwhile option if easy enough to implement.

**Comment by jzeevi** [comment]
Well, I have no good news to report. However, it’s from a test environment that I’ve not used before.

 

I loaded up 5b6. I have 2 SDR’s, and the  <http://www.ebay.com/itm/181566850426?_trksid=p2057872.m2749.l2649&ssPageName=STRK%3AMEBIDX%3AIT> RY835AI with a USB connection. I used airplane power or a power brick in the office/house.

 

This test was with my iPhone 6, not with my iPad. I believe that may have something to do with what I observed. I use Foreflight.

 

Prior to flight, I powered everything up and connected my iPhone to the stratux wifi. Saw traffic., WX, all seemed normal.

 

Took off, headed out for some flying and pretty quickly lost the wifi connection. The unit appeared to be working, but I couldn’t find the wifi on the phone anymore. So, I recycled the startux. Still no luck. I tried several times, but couldn’t get a connection. At that point, I needed to focus a bit on the flying.

 

I’ll try again this weekend, but with my iPad. Also, I know FF doesn’t do the AHRS but how in the world does port 40000 (I think that’s the back end port for the sim data (or is it the other one?) get disabled? Is that on the stratux side or the client (FF, iFly, etc) end? Or, is it a platform (tablet/phone) thing?

 

josef

**Comment by Nokomis449** [comment]
From a terminal, "sudo nano /etc/hostapd/hostapd.conf".  You can change the SSID and the channel. To save changes hit CTRL-X, then y, then ENTER.

**Comment by Ergonomicmike** [comment]
Can I do that in flight to try different channels? So far I only know how to SSH in via puTTY on my desktop with a wired Ethernet connection. I have Terminal as an app on my tablets (with BusyBox) if that helps. I know that Raspi Check basically was an SSH tool - but more like a UI than a Terminal. I guess what I'm asking is: Is there a free SSH Wi-Fi tool for Android?

**Comment by Axtel4** [comment]
I use wither JuiceSSH or vSSH.

>Is there a free SSH Wi-Fi tool for Android?

**Comment by Nokomis449** [comment]
JUICESSH here, too on Android.

**Comment by Ergonomicmike** [comment]
Tnx guys. Got Juice.

**Comment by Ergonomicmike** [comment]
For anyone having disconnect issues in flight, it might be a power pack related problem. I serendipitously discovered that my Stratux will shut down after a while with two SDR's when the CPU continuously runs at full bore. (I was testing the Performance governor over at https://github.com/cyoung/stratux/issues/222.) As the Stratux is now, it's using the ondemand governor, which doesn't run as hard and so doesn't stress the battery as much.

Although ondemand can run at full bore when necessary, it probably doesn't do it continuously in normal use, as Performance does. So Performance governor is the ultimate stress test, the ultimate worst case scenario. Naturally, if your power pack can't pass this test, then there might be times when ondemand runs full bore enough that it might cause the Stratux to drop out if you're using the KMASHI.

So if you're having disconnect issues, I suggest temporarily setting the governor to Performance (Instructions at http://raspberrypi.stackexchange.com/questions/9034/how-to-change-the-default-governor) to ensure that your power pack is not part of the problem. View it like a "burn in" test, to ensure all the components are up to the task. Might be a good "sanity test" in general for anyone's Stratux setup. (The governor will revert back to ondemand on reboot.)

**Comment by cyoung** [comment]
OK, I think we've fixed a few different issues here but there remain some - this issue discussion has too many different branches to work effectively on the remaining ones.

So for anyone that has the issue re-occur, please keep in mind the data points specific to your setup such that we can reproduce the issue. Then start a new issue for your specific case.

Some things that may be important:
1. WiFi signal level on the receiving device. Low signal means low data rate. There is some strange handling of slow WiFi clients in the WiFi dongle driver when the data rate exceeds the ability of the client to receive.
2. High data rates (check your UAT reception rate on the web interface, above 1500msg/min is "high").
3. Perhaps power issues as @Ergonomicmike.

So when it does re-occur, keep these things in mind and create a new issue using the issue format: https://github.com/cyoung/stratux/blob/master/CONTRIBUTING.md.

**Comment by Ergonomicmike** [comment]
Question: Does the web UI report msg/min? Is that what I'm seeing when it shows a number in the gas gauge? (I'm assuming it simply counts messages as they come in... although it does decay when I pull the antenna.) Regardless, if 1500 msgs/min can be a potential problem, then perhaps the Web UI could draw a red line (or equivalent) in the UAT Message gas gauge tool to warn users that there might be interconnect issues?

**Comment by cyoung** [comment]
It will just buffer the data above a certain level, that level was set to account for a decent WiFi signal. So these factors together compound the issue.

It's what you see on the web interface "status" (main) page. The number is the number of messages for UAT or 1090ES that you've received in the last minute.

**Comment by jzeevi** [comment]
FWIW I was running on continuous plane power and didn't see any lights go out or get a reboot sense. 

Josef

**Comment by cyoung** [comment]
No WiFi lockup on your setup with plane power then, @jzeevi?

**Comment by jzeevi** [comment]
Hard to say. The wifi was the only light I could not see....

But, i did get disconnected and stay disconnected from my iPhone. Not sure if that's a phone thing or a Stratux thing. I should be flying more this weekend and will take the iPad. Did my other report not come through?

**Comment by braddward** [comment]
Good news and bad news to report on this -

Bad news - broke the antenna on the stratux so I couldn't test it yesterday.
Good news - I passed my check ride so I'm official now.

I need to order new antennas instead of the cheapies that came with the
SDR's.

**Comment by JohnOCFII** [comment]
> Good news - I passed my check ride so I'm official now.

Congratulations!

**Comment by skypuppy** [comment]
Yes, HUGE contratulations!  Now, that smile will not go away for about a
week, so be advised.
